onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc

```diff
@@ -0,0 +1,476 @@
+  //   return size;
+  // This does exactly the same as the logic above
+  return (size + kDefaultByteAlignmentForNNAPI - 1) & ~(kDefaultByteAlignmentForNNAPI - 1);
+}
+
+void ModelBuilder::GetAllInitializers() {
+  for (const auto& tensor : model_proto_.graph().initializer()) {
+    initializers_.insert({tensor.name(), tensor});
+  }
+}
+
+void ModelBuilder::PreprocessIntializers() {
+  for (const auto& node : model_proto_.graph().node()) {
+    if (auto* opBuilder = GetOpBuilder(node)) {
+      opBuilder->AddInitializersToSkip(*this, node);
+    }
+  }
+}
+
+void ModelBuilder::RegisterInitializers() {
+  // First pass to get all the stats of the initializers
+  auto initializer_size = model_proto_.graph().initializer_size();
+  std::vector<std::tuple<uint32_t, size_t, size_t>> initializers(initializer_size);
+  size_t sizeAll = 0;
+
+  for (int i = 0; i < initializer_size; ++i) {
+    const auto& tensor = model_proto_.graph().initializer(i);
+    const auto& name = tensor.name();
+    if (Contains(skipped_initializers_, name))
+      continue;
+
+    Shape shape;
+    for (auto dim : tensor.dims()) {
+      shape.push_back(SafeInt<uint32_t>(dim));
+    }
+
+    Type type = Type::TENSOR_FLOAT32;
+    switch (tensor.data_type()) {
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:
+        type = Type::TENSOR_FLOAT32;
+        break;
+      default:
+        // TODO: support other type
+        ORT_THROW("The initializer of graph doesn't have valid type, name: " +
+                  name + " type: " + std::to_string(tensor.data_type()));
+        break;
+    }
+
+    OperandType operand_type(type, shape);
+    shaper_.AddShape(name, operand_type.dimensions);
+
+    auto index = AddNewOperand(name, operand_type);
+    const size_t size = operand_type.GetOperandBlobByteSize();
+    const size_t paddedSize = GetPaddedByteSize(size);
+    sizeAll += paddedSize;
+    initializers[i] = std::make_tuple(index, size, paddedSize);
+  }
+
+  // 2nd pass copies all the initializer data into NNAPI shared memory
+  nnapi_model_->mem_initializers_ =
+      std::make_unique<Model::NNMemory>(nnapi_, "mem_initializers_", sizeAll);
+
+  // 2nd pass to copy all the initializers into shared memory
+  size_t offset = 0;
+  for (int i = 0; i < initializer_size; ++i) {
+    const auto& tensor = model_proto_.graph().initializer(i);
+    if (Contains(skipped_initializers_, tensor.name()))
+      continue;
+
+    Index index;
+    size_t size, paddedSize;
+    std::tie(index, size, paddedSize) = initializers[i];
+    const char* src = nullptr;
+    if (tensor.data_type() == ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
+      src = tensor.float_data().empty()
+                ? tensor.raw_data().data()
+                : reinterpret_cast<const char*>(tensor.float_data().data());
+    } else {
+      ORT_THROW("The initializer of graph doesn't have valid type, name: " +
+                tensor.name() + " type: " + std::to_string(tensor.data_type()));
+    }
+
```

Comment:
Should this have resulted in it exiting early so a duplicate check isn't necessary?